tools/pubsys/src/aws/validate_ssm/results.rs

```diff
@@ -0,0 +1,681 @@
+//! The results module owns the reporting of SSM validation results.
+
+use crate::aws::ssm::ssm::Result;
+use aws_sdk_ssm::Region;
+use serde::ser::SerializeStruct;
+use serde::{Deserialize, Serialize};
+use std::collections::{HashMap, HashSet};
+use std::fmt::{self, Display};
+use std::str::FromStr;
+use tabled::{Table, Tabled};
+
+/// Represent the possible status of an SSM validation
+#[derive(Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
+pub enum SsmValidationResultStatus {
+    // The expected value was equal to the actual value
+    Correct,
+
+    // The expected value was different from the actual value
+    Incorrect,
+
+    // The parameter was expected but not included in the actual parameters
+    Missing,
+
+    // The parameter was present in the actual parameters but not expected
+    Unexpected,
+}
+
+impl Display for SsmValidationResultStatus {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::Correct => write!(f, "Correct"),
+            Self::Incorrect => write!(f, "Incorrect"),
+            Self::Missing => write!(f, "Missing"),
+            Self::Unexpected => write!(f, "Unexpected"),
+        }
+    }
+}
+
+impl FromStr for SsmValidationResultStatus {
+    type Err = super::Error;
+
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        match s {
+            "Correct" => Ok(Self::Correct),
+            "Incorrect" => Ok(Self::Incorrect),
+            "Missing" => Ok(Self::Missing),
+            "Unexpected" => Ok(Self::Unexpected),
+            filter => Err(Self::Err::InvalidStatusFilter {
+                filter: filter.to_string(),
+            }),
+        }
+    }
+}
+
+/// Represents a single SSM validation result
+#[derive(Debug, Eq, Hash, PartialEq, Tabled)]
+pub struct SsmValidationResult {
+    // The name of the parameter
+    pub(crate) name: String,
+
+    // The expected value of the parameter
+    #[tabled(display_with = "display_option")]
+    pub(crate) expected_value: Option<String>,
+
+    // The actual retrieved value of the parameter
+    #[tabled(display_with = "display_option")]
+    pub(crate) actual_value: Option<String>,
+
+    // The region the parameter resides in
+    pub(crate) region: Region,
+
+    // The ID of the AMI the parameter is associated with
+    #[tabled(display_with = "display_option")]
+    pub(crate) ami_id: Option<String>,
+
+    // The validation status of the parameter
+    pub(crate) status: SsmValidationResultStatus,
+}
+
+fn display_option(option: &Option<String>) -> &str {
+    match option {
+        Some(option) => option,
+        None => "N/A",
```

Comment:
Might be cleaner in the display?

```suggestion
        None => "",
```

Reply:
I feel like this is a matter of opinion. An empty string could suggest that the empty string was actually retrieved as value from SSM, which would make the parameter "Incorrect" instead of "Missing"/"Unexpected".